src/ethereum/cancun/vm/instructions/system.py

```diff
@@ -519,15 +520,18 @@ def selfdestruct(evm: Evm) -> None:
     # beneficiary).
     set_account_balance(evm.env.state, originator, U256(0))
 
-    # register account for deletion
-    evm.accounts_to_delete.add(originator)
+    # Only continue if the contract has been created in the same tx
+    if originator in evm.env.created_contracts:
 
-    # mark beneficiary as touched
-    if account_exists_and_is_empty(evm.env.state, beneficiary):
-        evm.touched_accounts.add(beneficiary)
+        # register account for deletion
+        evm.accounts_to_delete.add(originator)
 
-    # HALT the execution
-    evm.running = False
+        # mark beneficiary as touched
+        if account_exists_and_is_empty(evm.env.state, beneficiary):
+            evm.touched_accounts.add(beneficiary)
+
+        # HALT the execution
+        evm.running = False
```

Comment:
Hm, I think this'll also lead to an issue. If `evm.running` is not set to `False`, then you need to increment the program counter.

Reply:
Execution of the stack frame should end and return regardless of whether the contract was marked for deletion. So this should also be outside the `if` block.